pilot/pkg/proxy/envoy/v2/eds_sh_test.go

```diff
@@ -0,0 +1,324 @@
+// each network. It creates an instance of memory registry for each cluster and populate it
+// with Service, Instances and an ingress gateway service.
+// It then conducts an EDS query from each network expecting results to match the design of
+// the Split Horizon EDS - all local endpoints + endpoint per remote network that also has
+// endpoints for the service.
+func TestSplitHorizonEds(t *testing.T) {
+	initSplitHorizonTestEnv(t)
+
+	// Set up a cluster registry for network 1 with 1 instance for the service 'service5'
+	// Network has 1 gateway
+	initNetwork(1, []string{"159.122.219.1"}, 1)
+	// Set up a cluster registry for network 2 with 2 instances for the service 'service5'
+	// Network has 1 gateway
+	initNetwork(2, []string{"159.122.219.2"}, 2)
+	// Set up a cluster registry for network 3 with 3 instances for the service 'service5'
+	// Network has 2 gateways
+	initNetwork(3, []string{"159.122.219.3", "179.114.119.3"}, 3)
+	// Set up a cluster registry for network 4 with 4 instances for the service 'service5'
+	// but without any gateway
+	initNetwork(4, []string{}, 4)
+
+	// Update cache
+	pilotServer.EnvoyXdsServer.ClearCache()
+
+	// Verify that EDS from network1 will return 1 local endpoint with local VIP + 2 remote
+	// endpoints weighted accordingly with the IP of the ingress gateway.
+	verifySplitHorizonResponse(t, "network1", sidecarId("10.1.0.1", "app3"), ExpectedResults{
+		numOfNetworks:  3,
+		localEndpoints: []string{"10.1.0.1"},
+		remoteWeights:  map[string]uint32{"159.122.219.2": 2, "159.122.219.3": 3},
+	})
+
+	// Verify that EDS from network2 will return 2 local endpoints with local VIPs + 2 remote
+	// endpoints weighted accordingly with the IP of the ingress gateway.
+	verifySplitHorizonResponse(t, "network2", sidecarId("10.2.0.1", "app3"), ExpectedResults{
+		numOfNetworks:  3,
+		localEndpoints: []string{"10.2.0.1", "10.2.0.2"},
+		remoteWeights:  map[string]uint32{"159.122.219.1": 1, "159.122.219.3": 3},
+	})
+
+	// Verify that EDS from network3 will return 3 local endpoints with local VIPs + 2 remote
+	// endpoints weighted accordingly with the IP of the ingress gateway.
+	verifySplitHorizonResponse(t, "network3", sidecarId("10.3.0.1", "app3"), ExpectedResults{
+		numOfNetworks:  3,
+		localEndpoints: []string{"10.3.0.1", "10.3.0.2", "10.3.0.3"},
+		remoteWeights:  map[string]uint32{"159.122.219.1": 1, "159.122.219.2": 2},
+	})
+
+	verifySplitHorizonResponse(t, "network4", sidecarId("10.4.0.1", "app3"), ExpectedResults{
+		numOfNetworks:  4,
+		localEndpoints: []string{"10.4.0.1", "10.4.0.2", "10.4.0.3", "10.4.0.4"},
+		remoteWeights:  map[string]uint32{"159.122.219.1": 1, "159.122.219.2": 2, "159.122.219.3": 3},
+	})
+}
+
+// Tests whether an EDS response from the provided network matches the expected results
+func verifySplitHorizonResponse(t *testing.T, network string, sidecarId string, expected ExpectedResults) {
+	edsstr, err := connectADS(util.MockPilotGrpcAddr)
```

Comment:
call `t.Helper()` as the first thing in this method so that if/when the tests fails in this method the failure is reported relative to the caller (so you can see which case above actually failed)